src/coordinate_systems/XYZT.jl

```diff
@@ -0,0 +1,174 @@
+"""
+
+    XYZT <: AbstractCoordinateSystem
+
+Cartesian coordinate system for four-vectors. Using this requires the implementation of the following interface functions:
+
+```julia
+LorentzVectorBase.x(::CustomFourVector)
+LorentzVectorBase.y(::CustomFourVector)
+LorentzVectorBase.z(::CustomFourVector)
+LorentzVectorBase.t(::CustomFourVector)
+```
+
+"""
+struct XYZT <: AbstractCoordinateSystem end
+coordinate_names(::XYZT) = (:x, :y, :z, :t)
+
+####
+# Interface functions
+####
+
+function t end
+
+function x end
+
+function y end
+
+function z end
+
+####
+# derived components
+####
+
+@inline function spatial_magnitude2(::XYZT, mom)
+  return x(mom)^2 + y(mom)^2 + z(mom)^2
+end
+
+@inline function spatial_magnitude(cs::XYZT, mom)
+  return sqrt(spatial_magnitude2(mom))
+end
+
+@inline function mass2(::XYZT, mom)
+  return t(mom)^2 - x(mom)^2 - y(mom)^2 - z(mom)^2
+end
+
+@inline function mass(::XYZT, mom)
+  m2 = mass2(mom)
+  if m2 < zero(m2)
+    # Think about including this warning, maybe optional with a global PRINT_WARNINGS switch.
+    #@warn("The square of the invariant mass (m2=P*P) is negative. The value -sqrt(-m2) is returned.")
+    return -sqrt(-m2)
+  else
+    return sqrt(m2)
+  end
+end
+
+@inline function boost_beta(::XYZT, mom)
+  en = t(mom)
+  rho = spatial_magnitude(mom)
+  if !iszero(en)
+    return rho / en
+  elseif iszero(rho)
+    return zero(rho)
+  else
+    throw(
+      ArgumentError(
+        "there is no beta for a four-vector with vanishing time/energy component"
+      ),
+    )
+  end
+end
+
+@inline function boost_gamma(::XYZT, mom)
+  beta = boost_beta(mom)
+  return inv(sqrt(one(beta) - beta^2))
+end
+
+########################
+# transverse coordinates
+########################
+@inline function pt2(::XYZT, mom)
+  return x(mom)^2 + y(mom)^2
+end
+
+@inline function pt(::XYZT, mom)
+  return sqrt(pt2(mom))
+end
+
+@inline function mt2(::XYZT, mom)
+  return t(mom)^2 - z(mom)^2
+end
+
+function mt(::XYZT, mom)
+  mT2 = mt2(mom)
+  if mT2 < zero(mT2)
+    # add optional waring: negative transverse mass -> -sqrt(-mT2) is returned.
+    -sqrt(-mT2)
+  else
+    sqrt(mT2)
+  end
+end
+
+@inline function rapidity(::XYZT, mom)
+  en = t(mom)
+  zcomp = z(mom)
+  return 0.5 * log((en + zcomp) / (en - zcomp))
+end
+
+function eta(::XYZT, mom)
+  cth = cos_theta(mom)
+
+  if cth^2 < one(cth)
+    return -0.5 * log((1 - cth) / (1 + cth))
+  end
+
+  zcomp = z(mom)
+  if iszero(zcomp)
+    return zero(zcomp)
+  end
+
+  @warn "Pseudorapidity (η): transverse momentum is zero! return +/- 10e10"
```

Comment:
I don't understand this message, it's not zero right? Why does it say it's zero.

And, why would eta ever be infinite? I thought by definition it can't be infinite

Reply:
The only possible way to reach this point is `cos_theta^2 == 1` and `pz!=0`. Therefore, `pT = sin_theta*rho==0`. 

In this case, `eta == -Inf` if `cth == -1` and `eta == Inf` if `cth==1`. But you are right, the last guarding number is wrong, fixed it in 0eee94e. 

Btw: this implementation was originally copied from [here](https://root.cern.ch/doc/master/classTVector3.html#a60284c6bd7ac08940b129b514b135288).